### PR TITLE
ARROW-7498: [Dataset] Rename core classes before stable API

### PR DESCRIPTION
**Pull request by fsaintjacques**
- Rename DataFragment to Fragment
- Rename DataSource to Source
- Rename *Discovery to *Manifest
- Rename PartitionScheme to Partitioner
- Rename Simple* to InMemory*
- Rename SchemaPartitioner to DirectoryPartitioner

### CODE REVIEW

**Comment by fsaintjacques** [comment]
Get your favorite pantone color ready, now is the time to paint the bikeshed!

**Comment by pitrou** [comment]
I don't think what you call "partitioner" actually partitions data, does it?

**Comment by fsaintjacques** [comment]
@pitrou correct, @jorisvandenbossche noted this in JIRA.

**Comment by jorisvandenbossche** [comment]
So le's go with "Partitioning" then?

**Comment by jorisvandenbossche** [comment]
What is the reasoning behind the *Discovery to *Manifest rename? 

For me, a "manifest" sounds something as a full description (listing of all the partitions, all files and their schema's), while the current FileSystemDataSourceDiscovery is only an incomplete description (eg only a directory name) and it *discovers* the full description (all files and their schema's in the directory, the partitioning of the directory)

File: cpp/src/arrow/dataset/dataset.cc [hunk 1]
```diff
@@ -29,15 +29,15 @@
 namespace arrow {
 namespace dataset {
 
-DataFragment::DataFragment(std::shared_ptr<ScanOptions> scan_options)
+Fragment::Fragment(std::shared_ptr<ScanOptions> scan_options)
     : scan_options_(std::move(scan_options)), partition_expression_(scalar(true)) {}
 
-SimpleDataFragment::SimpleDataFragment(
+InMemoryFragment::InMemoryFragment(
```
Comment (kszucs) [inline cpp/src/arrow/dataset/dataset.cc]: +1

**Comment by kszucs** [comment]
@fsaintjacques how about a bit different approach for Discovery/Manifest.

### Proposal 

Remove the `*Manifest` postfix from the classes, and add `Materialized*` (or something similar/shorter) prefix for the counterpart classes, e.g.:

```
FileSystemSourceManifest -> FileSystemSource
FileSystemSource -> MaterializedFileSystemSource
```

### Reasoning

Correct me if I'm wrong, but the end user will most likely interact with the `*Manifest` classes, because the user don't want to do the discovery manually, the user just want to describe the datasource. 

If I recall correctly you were planning a DatasetManifest (DatasetDiscovery) counterpart for Dataset as well, mainly to encapsulate the boilerplate required to materialize/discover and unify the schemas of the datasources. So I'd rather imagine the following:

```python
local = FileSystemSource(localfs, ...). # currently FileSystemSourceManifest
s3 = FileSystemSource(s3fs, ...)   # currently FileSystemSourceManifest
ds = Dataset([local, s3])  # currently this would be DatasetManifest
```

The "plain" classes would be used for describe a dataset, then the `Dataset[Manifest]` would handle the required discovery logic resulting the materialized objects having all the required metadata (file paths, schemas etc) in memory.

**Comment by kszucs** [comment]
> So le's go with "Partitioning" then?

I like Partitioning more than Partitioner (if we can't figure out a shorter alternative :).

**Comment by fsaintjacques** [comment]
I don't like the "Materialized" keyword, it's an implementation detail that leaks. How about we just name Discovery/Manifest to Factory. While boring, it's obvious to everyone who look at the class.

**Comment by kszucs** [comment]
@fsaintjacques Aren't the classes without the Discovery prefixes (DataFragment, DataSource, Dataset, PartitionScheme) implementation detail by themselves?

**Comment by kszucs** [comment]
@fsaintjacques shouldn't we call Fragment as Partition?

```
  /// \brief An expression which evaluates to true for all data viewed by this
  /// DataFragment. May be null, which indicates no information is available.
  const std::shared_ptr<Expression>& partition_expression() const {
    return partition_expression_;
  }
```

**Comment by jorisvandenbossche** [comment]
> shouldn't we call Fragment as Partition?

There can be multiple Fragments in a Partition.

(at least in my head when thinking about fragments as single files and a "partition" a subdirectory with a given set of values for its partition keys)

**Comment by kszucs** [comment]
A FileDataFragment is [constructed from a single file](https://github.com/apache/arrow/blob/8bd27fe97ec51feb627c56492b6a8736949cc4a0/cpp/src/arrow/dataset/file_base.h#L156).

**Comment by jorisvandenbossche** [comment]
So a summary of the renaming in the current state of this PR:

* DataFragment -> Fragment
* DataSource -> Source
* Simple* -> InMemory*
* *Discovery ->  *Factory
* *PartitionScheme -> Partitioning
* SchemaPartitioning -> DirectoryPartitioning

Are we fine with this?

I personally find "Directory" still a bit strange, as the HivePartitioning is also directory-based.

**Comment by fsaintjacques** [comment]
I'm fine to change DirectoryPartitioning to something else. In the ticket:

-StackPartitioner
-LevelPartitioner
-HierarchyPartitioner
-OrderedPartitioner
-DirectoryPartitioner

**Comment by jorisvandenbossche** [comment]
Personally, I am not sure if I find any of those alternatives better than `SchemaPartitioning` .. (and now that the "Scheme" part is gone, you also don't have the Schema/Scheme confusing/repetition)

**Pull request closed without merging**